File: src/main/java/com/enderdash/kestara/webrtc/PeerConnection.java
package com.enderdash.kestara.webrtc;

import com.enderdash.kestara.webrtc.internal.NativeBindings;
import com.enderdash.kestara.webrtc.internal.NativeEvent;
import com.enderdash.kestara.webrtc.internal.SerialExecutor;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;

/** A WebRTC peer connection for DataChannel negotiation and transport. */
public final class PeerConnection implements AutoCloseable {
    private static final Consumer<SessionDescription> NOOP_DESCRIPTION = ignored -> {};
    private static final Consumer<IceCandidate> NOOP_CANDIDATE = ignored -> {};
    private static final Consumer<DataChannel> NOOP_DATA_CHANNEL = ignored -> {};
    private static final Consumer<PeerConnectionState> NOOP_STATE = ignored -> {};
    private static final Consumer<IceConnectionState> NOOP_ICE_STATE = ignored -> {};
    private static final Consumer<IceGatheringState> NOOP_GATHERING_STATE = ignored -> {};

    private final long handle;
    private final long operationTimeoutMillis;
    private final SerialExecutor callbacks;
    private final Map<Long, DataChannel> dataChannels = new ConcurrentHashMap<>();
    private final AtomicBoolean closed = new AtomicBoolean();
    private volatile PeerConnectionState state = PeerConnectionState.NEW;
    private volatile IceConnectionState iceConnectionState = IceConnectionState.NEW;
    private volatile IceGatheringState iceGatheringState = IceGatheringState.NEW;
    private volatile Consumer<SessionDescription> onLocalDescription = NOOP_DESCRIPTION;
    private volatile Consumer<IceCandidate> onLocalCandidate = NOOP_CANDIDATE;
    private volatile Consumer<DataChannel> onDataChannel = NOOP_DATA_CHANNEL;
    private volatile Consumer<PeerConnectionState> onStateChange = NOOP_STATE;
    private volatile Consumer<IceConnectionState> onIceConnectionStateChange = NOOP_ICE_STATE;
    private volatile Consumer<IceGatheringState> onIceGatheringStateChange = NOOP_GATHERING_STATE;

    private PeerConnection(long handle, PeerConnectionConfiguration configuration) {
        this.handle = handle;
        this.operationTimeoutMillis = configuration.operationTimeoutMillis();
        this.callbacks = new SerialExecutor(configuration.callbackExecutor());
    }

    /**
     * Creates a peer connection and starts the shared native runtime when necessary.
     *
     * @param configuration the peer configuration
     * @return the peer connection
     */
    public static PeerConnection create(PeerConnectionConfiguration configuration) {
        Objects.requireNonNull(configuration, "configuration");
        NativeConfiguration nativeConfiguration = NativeConfiguration.from(configuration);
        long handle = NativeBindings.createPeer(
                nativeConfiguration.urls,
                nativeConfiguration.usernames,
                nativeConfiguration.credentials,
                configuration.minPort(),
                configuration.maxPort(),
                configuration.iceTransportPolicy().ordinal(),
                configuration.dataChannelSendBufferLimit(),
                configuration.operationTimeoutMillis());
        try {
            PeerConnection peer = new PeerConnection(handle, configuration);
            NativeEventDispatcher.register(handle, peer);
            return peer;
        } catch (RuntimeException | Error error) {
            try {
                NativeBindings.closePeer(handle, configuration.operationTimeoutMillis());
            } catch (RuntimeException closeError) {
                error.addSuppressed(closeError);
            }
            throw error;
        }
    }

    /**
     * Returns the aggregate peer state.
     *
     * @return the peer state
     */
    public PeerConnectionState state() {
        return state;
    }

    /**
     * Returns the ICE connectivity state.
     *
     * @return the ICE state
     */
    public IceConnectionState iceConnectionState() {
        return iceConnectionState;
    }

    /**
     * Returns the local candidate-gathering state.
     *
     * @return the gathering state
     */
    public IceGatheringState iceGatheringState() {
        return iceGatheringState;
    }

    /**
     * Sets the callback for locally applied descriptions.
     *
     * @param callback the callback
     */
    public void onLocalDescription(Consumer<SessionDescription> callback) {
        onLocalDescription = Objects.requireNonNull(callback, "callback");
    }

    /**
     * Sets the callback for trickled local ICE candidates.
     *
     * @param callback the callback
     */
    public void onLocalCandidate(Consumer<IceCandidate> callback) {
        onLocalCandidate = Objects.requireNonNull(callback, "callback");
    }

    /**
     * Sets the callback for channels created by the remote peer.
     *
     * @param callback the callback
     */
    public void onDataChannel(Consumer<DataChannel> callback) {
        onDataChannel = Objects.requireNonNull(callback, "callback");
    }

    /**
     * Sets the callback for aggregate peer-state changes.
     *
     * @param callback the callback
     */
    public void onStateChange(Consumer<PeerConnectionState> callback) {
        onStateChange = Objects.requireNonNull(callback, "callback");
    }

    /**
     * Sets the callback for ICE connectivity changes.
     *
     * @param callback the callback
     */
    public void onIceConnectionStateChange(Consumer<IceConnectionState> callback) {
        onIceConnectionStateChange = Objects.requireNonNull(callback, "callback");
    }

    /**
     * Sets the callback for local candidate-gathering changes.
     *
     * @param callback the callback
     */
    public void onIceGatheringStateChange(Consumer<IceGatheringState> callback) {
        onIceGatheringStateChange = Objects.requireNonNull(callback, "callback");
    }

    /**
     * Creates an SDP offer without applying it.
     *
     * @return the offer
     */
    public SessionDescription createOffer() {
        requireOpen();
        return new SessionDescription(
                NativeBindings.createDescription(handle, SessionDescriptionType.OFFER.ordinal()),
                SessionDescriptionType.OFFER);
    }

    /**
     * Creates an SDP answer without applying it.
     *
     * @return the answer
     */
    public SessionDescription createAnswer() {
        requireOpen();
        return new SessionDescription(
                NativeBindings.createDescription(handle, SessionDescriptionType.ANSWER.ordinal()),
                SessionDescriptionType.ANSWER);
    }

    /**
     * Applies a local session description and reports it to the local-description callback.
     *
     * @param description the description
     */
    public void setLocalDescription(SessionDescription description) {
        Objects.requireNonNull(description, "description");
        requireOpen();
        NativeBindings.setLocalDescription(
                handle, description.sdp(), description.type().ordinal(), operationTimeoutMillis);
        callbacks.execute(() -> onLocalDescription.accept(description));
    }

    /**
     * Creates and applies an offer or answer.
     *
     * @param type the description type
     * @return the applied description
     */
    public SessionDescription setLocalDescription(SessionDescriptionType type) {
        Objects.requireNonNull(type, "type");
        SessionDescription description = switch (type) {
            case OFFER -> createOffer();
            case ANSWER -> createAnswer();
            case PRANSWER, ROLLBACK -> throw new UnsupportedOperationException(
                    "Automatic local description does not support " + type);
        };
        setLocalDescription(description);
        return description;
    }

    /**
     * Applies a remote session description.
     *
     * @param description the description received through signaling
     */
    public void setRemoteDescription(SessionDescription description) {
        Objects.requireNonNull(description, "description");
        requireOpen();
        NativeBindings.setRemoteDescription(
                handle, description.sdp(), description.type().ordinal(), operationTimeoutMillis);
    }

    /**
     * Adds a remote trickle ICE candidate.
     *
     * @param candidate the candidate received through signaling
     */
    public void addIceCandidate(IceCandidate candidate) {
        Objects.requireNonNull(candidate, "candidate");
        requireOpen();
        NativeBindings.addIceCandidate(
                handle,
                candidate.candidate(),
                candidate.sdpMid(),
                candidate.sdpMLineIndex() == null ? -1 : candidate.sdpMLineIndex(),
                operationTimeoutMillis);
    }

    /**
     * Creates an ordered and reliable DataChannel.
     *
     * @param label the channel label
     * @return the channel
     */
    public DataChannel createDataChannel(String label) {
        return createDataChannel(label, DataChannelOptions.DEFAULT);
    }

    /**
     * Creates a DataChannel.
     *
     * @param label the channel label
     * @param options the channel options
     * @return the channel
     */
    public DataChannel createDataChannel(String label, DataChannelOptions options) {
        Objects.requireNonNull(label, "label");
        Objects.requireNonNull(options, "options");
        requireOpen();
        long channelHandle = NativeBindings.createDataChannel(
                handle,
                label,
                options.ordered(),
                optionalUnsigned16(options.maxPacketLifeTime()),
                optionalUnsigned16(options.maxRetransmits()),
                options.protocol(),
                optionalUnsigned16(options.negotiatedId()),
                operationTimeoutMillis);
        DataChannel channel =
                new DataChannel(
                        channelHandle,
                        label,
                        options.protocol(),
                        options.ordered(),
                        callbacks,
                        false);
        dataChannels.put(channelHandle, channel);
        return channel;
    }

    /** Closes the peer and all owned DataChannels. */
    @Override
    public void close() {
        if (!closed.compareAndSet(false, true)) {
            return;
        }
        NativeEventDispatcher.unregister(handle);
        for (DataChannel channel : dataChannels.values()) {
            channel.markClosed();
        }
        dataChannels.clear();
        try {
            NativeBindings.closePeer(handle, operationTimeoutMillis);
        } finally {
            state = PeerConnectionState.CLOSED;
        }
    }

    void closeForShutdown() {
        if (!closed.compareAndSet(false, true)) {
            return;
        }
        NativeEventDispatcher.unregister(handle);
        for (DataChannel channel : dataChannels.values()) {
            channel.markClosed();
        }
        dataChannels.clear();
        state = PeerConnectionState.CLOSED;
    }

    void handleNativeEvent(NativeEvent event) {
        switch (event.kind()) {
            case NativeBindings.EVENT_LOCAL_CANDIDATE -> {
                Integer lineIndex = event.number() < 0 ? null : event.number();
                IceCandidate candidate =
                        new IceCandidate(event.text(), event.secondaryText(), lineIndex);
                callbacks.execute(() -> onLocalCandidate.accept(candidate));
            }
            case NativeBindings.EVENT_PEER_STATE -> {
                PeerConnectionState newState = enumValue(
                        PeerConnectionState.values(), event.number(), "peer connection state");
                state = newState;
                callbacks.execute(() -> onStateChange.accept(newState));
            }
            case NativeBindings.EVENT_ICE_CONNECTION_STATE -> {
                IceConnectionState newState = enumValue(
                        IceConnectionState.values(), event.number(), "ICE connection state");
                iceConnectionState = newState;
                callbacks.execute(() -> onIceConnectionStateChange.accept(newState));
            }
            case NativeBindings.EVENT_ICE_GATHERING_STATE -> {
                IceGatheringState newState = enumValue(
                        IceGatheringState.values(), event.number(), "ICE gathering state");
                iceGatheringState = newState;
                callbacks.execute(() -> onIceGatheringStateChange.accept(newState));
            }
            case NativeBindings.EVENT_DATA_CHANNEL -> {
                boolean ordered = (event.number() & 1) != 0;
                boolean initiallyOpen = (event.number() & 2) != 0;
                DataChannel channel = new DataChannel(
                        event.channelHandle(),
                        event.text() == null ? "" : event.text(),
                        event.secondaryText() == null ? "" : event.secondaryText(),
                        ordered,
                        callbacks,
                        initiallyOpen);
                dataChannels.put(event.channelHandle(), channel);
                callbacks.execute(() -> {
                    onDataChannel.accept(channel);
                    if (initiallyOpen) {
                        channel.notifyOpen();
                    }
                });
            }
            case NativeBindings.EVENT_DATA_CHANNEL_OPEN,
                    NativeBindings.EVENT_DATA_CHANNEL_CLOSING,
                    NativeBindings.EVENT_DATA_CHANNEL_CLOSED,
                    NativeBindings.EVENT_DATA_CHANNEL_ERROR,
                    NativeBindings.EVENT_DATA_CHANNEL_TEXT,
                    NativeBindings.EVENT_DATA_CHANNEL_BINARY -> {
                DataChannel channel = dataChannels.get(event.channelHandle());
                if (channel != null) {
                    channel.handleNativeEvent(event);
                    if (event.kind() == NativeBindings.EVENT_DATA_CHANNEL_CLOSED) {
                        dataChannels.remove(event.channelHandle(), channel);
                    }
                }
            }
            default -> throw new IllegalArgumentException("Unsupported peer event: " + event.kind());
        }
    }

    private void requireOpen() {
        if (closed.get()) {
            throw new IllegalStateException("PeerConnection is closed");
        }
    }

    private static int optionalUnsigned16(Integer value) {
        return value == null ? -1 : value;
    }

    private static <T> T enumValue(T[] values, int ordinal, String name) {
        if (ordinal < 0 || ordinal >= values.length) {
            throw new IllegalArgumentException("Unknown " + name + ": " + ordinal);
        }
        return values[ordinal];
    }

    private record NativeConfiguration(String[] urls, String[] usernames, String[] credentials) {
        private static NativeConfiguration from(PeerConnectionConfiguration configuration) {
            List<String> urls = new ArrayList<>();
            List<String> usernames = new ArrayList<>();
            List<String> credentials = new ArrayList<>();
            for (IceServer server : configuration.iceServers()) {
                for (String url : server.urls()) {
                    urls.add(url);
                    usernames.add(server.username());
                    credentials.add(server.credential());
                }
            }
            return new NativeConfiguration(
                    urls.toArray(String[]::new),
                    usernames.toArray(String[]::new),
                    credentials.toArray(String[]::new));
        }
    }
}

File: src/commonMain/kotlin/com/enderdash/kestara/webrtc/PeerConnection.kt
package com.enderdash.kestara.webrtc

import com.enderdash.kestara.webrtc.internal.NativeEvent
import com.enderdash.kestara.webrtc.internal.NativeEventKind
import kotlinx.atomicfu.atomic
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlin.time.Duration

public class PeerConnection internal constructor(
  private val runtime: WebRtcRuntime,
  private val handle: Long,
  configuration: PeerConnectionConfiguration,
) {
  private val timeout: Duration = configuration.operationTimeout
  private val receiveQueueCapacity: Int = configuration.sctpOptions.receiveQueueCapacity
  private val channelsMutex = Mutex()
  private val channels = mutableMapOf<Long, DataChannel>()
  private val channelCount = atomic(0)
  private val queuedChannelEvents = mutableMapOf<Long, MutableList<NativeEvent>>()
  private val closing = atomic(false)
  private val closed = atomic(false)
  private val closeCompletion = CompletableDeferred<Unit>()
  private val mutableState = MutableStateFlow(PeerConnectionState.NEW)
  private val mutableIceConnectionState = MutableStateFlow(IceConnectionState.NEW)
  private val mutableIceGatheringState = MutableStateFlow(IceGatheringState.NEW)
  private val mutableSignalingState = MutableStateFlow(SignalingState.STABLE)
  private val localDescriptionChannel = Channel<SessionDescription>(Channel.BUFFERED)
  private val localCandidateChannel = Channel<IceCandidate>(Channel.BUFFERED)
  private val incomingDataChannelChannel = Channel<DataChannel>(Channel.BUFFERED)
  private val negotiationNeededChannel = Channel<Unit>(Channel.CONFLATED)

  public val state: StateFlow<PeerConnectionState> = mutableState.asStateFlow()
  public val iceConnectionState: StateFlow<IceConnectionState> = mutableIceConnectionState.asStateFlow()
  public val iceGatheringState: StateFlow<IceGatheringState> = mutableIceGatheringState.asStateFlow()
  public val signalingState: StateFlow<SignalingState> = mutableSignalingState.asStateFlow()
  public val localDescriptions: Flow<SessionDescription> = localDescriptionChannel.receiveAsFlow()
  public val localCandidates: Flow<IceCandidate> = localCandidateChannel.receiveAsFlow()
  public val incomingDataChannels: Flow<DataChannel> = incomingDataChannelChannel.receiveAsFlow()
  public val negotiationNeeded: Flow<Unit> = negotiationNeededChannel.receiveAsFlow()

  public suspend fun createOffer(): SessionDescription {
    requireOpen()
    return SessionDescription(
      runtime.createDescription(handle, SessionDescriptionType.OFFER, timeout),
      SessionDescriptionType.OFFER,
    )
  }

  public suspend fun createAnswer(): SessionDescription {
    requireOpen()
    return SessionDescription(
      runtime.createDescription(handle, SessionDescriptionType.ANSWER, timeout),
      SessionDescriptionType.ANSWER,
    )
  }

  public suspend fun setLocalDescription(description: SessionDescription) {
    requireOpen()
    runtime.setLocalDescription(handle, description, timeout)
    localDescriptionChannel.send(description)
  }

  public suspend fun createAndSetLocalDescription(type: SessionDescriptionType): SessionDescription {
    val description = when (type) {
      SessionDescriptionType.OFFER -> createOffer()
      SessionDescriptionType.ANSWER -> createAnswer()
      SessionDescriptionType.PRANSWER,
      SessionDescriptionType.ROLLBACK,
      -> throw UnsupportedOperationException("Automatic local description does not support $type")
    }
    setLocalDescription(description)
    return description
  }

  public suspend fun setRemoteDescription(description: SessionDescription) {
    requireOpen()
    runtime.setRemoteDescription(handle, description, timeout)
  }

  public suspend fun addIceCandidate(candidate: IceCandidate) {
    requireOpen()
    runtime.addIceCandidate(handle, candidate, timeout)
  }

  public suspend fun restartIce() {
    requireOpen()
    runtime.restartIce(handle, timeout)
  }

  public suspend fun setConfiguration(configuration: PeerConnectionConfiguration) {
    requireOpen()
    runtime.setConfiguration(handle, configuration, timeout)
  }

  public suspend fun getStats(): PeerConnectionStats {
    requireOpen()
    return runtime.getStats(handle, timeout)
  }

  public suspend fun selectedIceCandidatePair(): IceCandidatePairStats? =
    getStats().transport.selectedCandidatePair

  public suspend fun createDataChannel(
    label: String,
    options: DataChannelOptions = DataChannelOptions(),
  ): DataChannel {
    requireOpen()
    val registration = runtime.createDataChannel(handle, label, options, timeout)
    val channel = DataChannel(
      runtime = runtime,
      handle = registration.handle,
      id = registration.id,
      label = label,
      protocol = options.protocol,
      ordered = options.ordered,
      negotiated = options.negotiatedId != null,
      maxPacketLifeTime = options.maxPacketLifeTime,
      maxRetransmits = options.maxRetransmits,
      receiveQueueCapacity = receiveQueueCapacity,
      timeout = timeout,
      initiallyOpen = false,
      onTerminal = { channelHandle -> unregisterDataChannel(channelHandle) },
    )
    val queued = channelsMutex.withLock {
      channels[registration.handle] = channel
      channelCount.value = channels.size
      queuedChannelEvents.remove(registration.handle).orEmpty()
    }
    queued.forEach { channel.handleNativeEvent(it) }
    return channel
  }

  public suspend fun close() {
    if (closing.compareAndSet(expect = false, update = true)) {
      try {
        runtime.closePeer(handle, timeout)
        markClosed(runtimeShutdown = false)
        closeCompletion.complete(Unit)
      } catch (error: Throwable) {
        markClosed(runtimeShutdown = false)
        closeCompletion.completeExceptionally(error)
      }
    }
    closeCompletion.await()
  }

  internal suspend fun handleNativeEvent(event: NativeEvent) {
    when (event.kind) {
      NativeEventKind.LOCAL_CANDIDATE -> localCandidateChannel.send(
        IceCandidate(
          candidate = event.text.orEmpty(),
          sdpMid = event.secondaryText,
          sdpMLineIndex = event.number.takeIf { it >= 0 },
        ),
      )
      NativeEventKind.PEER_STATE -> mutableState.value = enumValue(event.number, "peer connection state")
      NativeEventKind.ICE_CONNECTION_STATE -> {
        mutableIceConnectionState.value = enumValue(event.number, "ICE connection state")
      }
      NativeEventKind.ICE_GATHERING_STATE -> {
        mutableIceGatheringState.value = enumValue(event.number, "ICE gathering state")
      }
      NativeEventKind.SIGNALING_STATE -> {
        mutableSignalingState.value = enumValue(event.number, "signaling state")
      }
      NativeEventKind.NEGOTIATION_NEEDED -> negotiationNeededChannel.send(Unit)
      NativeEventKind.DATA_CHANNEL -> registerRemoteDataChannel(event)
      NativeEventKind.DATA_CHANNEL_OPEN,
      NativeEventKind.DATA_CHANNEL_CLOSING,
      NativeEventKind.DATA_CHANNEL_CLOSED,
      NativeEventKind.DATA_CHANNEL_ERROR,
      NativeEventKind.DATA_CHANNEL_TEXT,
      NativeEventKind.DATA_CHANNEL_BINARY,
      NativeEventKind.DATA_CHANNEL_BUFFERED_AMOUNT_LOW,
      NativeEventKind.DATA_CHANNEL_BUFFERED_AMOUNT_HIGH,
      -> routeDataChannelEvent(event)
      else -> throw WebRtcException("Unsupported peer event: ${event.kind}")
    }
  }

  internal fun dataChannelCount(): Int = channelCount.value

  internal suspend fun closeForRuntimeShutdown() {
    markClosed(runtimeShutdown = true)
    closeCompletion.complete(Unit)
  }

  private suspend fun registerRemoteDataChannel(event: NativeEvent) {
    val metadata = decodeChannelMetadata(event.data)
    val ordered = event.number and 1 != 0
    val initiallyOpen = event.number and 2 != 0
    val id = event.number ushr 2
    val channel = DataChannel(
      runtime = runtime,
      handle = event.channelHandle,
      id = id,
      label = event.text.orEmpty(),
      protocol = event.secondaryText.orEmpty(),
      ordered = ordered,
      negotiated = metadata.negotiated,
      maxPacketLifeTime = metadata.maxPacketLifeTime,
      maxRetransmits = metadata.maxRetransmits,
      receiveQueueCapacity = receiveQueueCapacity,
      timeout = timeout,
      initiallyOpen = initiallyOpen,
      onTerminal = { channelHandle -> unregisterDataChannel(channelHandle) },
    )
    val queued = channelsMutex.withLock {
      channels[event.channelHandle] = channel
      channelCount.value = channels.size
      queuedChannelEvents.remove(event.channelHandle).orEmpty()
    }
    incomingDataChannelChannel.send(channel)
    queued.forEach { channel.handleNativeEvent(it) }
    if (initiallyOpen) channel.notifyOpen()
  }

  private suspend fun routeDataChannelEvent(event: NativeEvent) {
    val channel = channelsMutex.withLock {
      channels[event.channelHandle] ?: run {
        queuedChannelEvents.getOrPut(event.channelHandle, ::mutableListOf).add(event)
        null
      }
    }
    channel?.handleNativeEvent(event)
  }

  private suspend fun unregisterDataChannel(channelHandle: Long) {
    channelsMutex.withLock {
      channels.remove(channelHandle)
      channelCount.value = channels.size
    }
  }

  private suspend fun markClosed(runtimeShutdown: Boolean) {
    if (!closed.compareAndSet(expect = false, update = true)) return
    closing.value = true
    val ownedChannels = channelsMutex.withLock {
      val copy = channels.values.toList()
      channels.clear()
      channelCount.value = 0
      queuedChannelEvents.clear()
      copy
    }
    ownedChannels.forEach { it.markClosed(runtimeShutdown) }
    mutableState.value = PeerConnectionState.CLOSED
    mutableSignalingState.value = SignalingState.CLOSED
    localDescriptionChannel.close()
    localCandidateChannel.close()
    incomingDataChannelChannel.close()
    negotiationNeededChannel.close()
    runtime.unregisterPeer(handle, this)
  }

  private fun requireOpen() {
    check(!closing.value && !closed.value) { "PeerConnection is closed" }
  }

  private inline fun <reified T : Enum<T>> enumValue(ordinal: Int, name: String): T =
    enumValues<T>().getOrNull(ordinal)
      ?: throw WebRtcException("Unknown $name: $ordinal")

  private data class ChannelMetadata(
    val negotiated: Boolean,
    val maxPacketLifeTime: Int?,
    val maxRetransmits: Int?,
  )

  private companion object {
    fun decodeChannelMetadata(data: ByteArray?): ChannelMetadata {
      require(data != null && data.size == 9) { "Invalid native DataChannel metadata" }
      fun intAt(offset: Int): Int =
        (data[offset].toInt() and 0xff) or
          ((data[offset + 1].toInt() and 0xff) shl 8) or
          ((data[offset + 2].toInt() and 0xff) shl 16) or
          ((data[offset + 3].toInt() and 0xff) shl 24)
      val lifetime = intAt(1)
      val retransmits = intAt(5)
      return ChannelMetadata(
        negotiated = data[0].toInt() != 0,
        maxPacketLifeTime = lifetime.takeIf { it >= 0 },
        maxRetransmits = retransmits.takeIf { it >= 0 },
      )
    }
  }
}
